File: app/src/main/java/com/narsha/wave_android/view/fragment/main/MainFragment.java
package com.narsha.wave_android.view.fragment.main;

import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.NavController;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.narsha.wave_android.R;
import com.narsha.wave_android.data.response.music.RecommendList;
import com.narsha.wave_android.data.response.music.PlayList;
import com.narsha.wave_android.data.slider.SliderItem;
import com.narsha.wave_android.network.Server;
import com.narsha.wave_android.view.adapter.imageslider.MainImageSlider;
import com.narsha.wave_android.view.adapter.recyclerview.MainSongAdapter;
import com.narsha.wave_android.viewmodel.MainViewModel;
import com.smarteist.autoimageslider.IndicatorView.animation.type.IndicatorAnimationType;
import com.smarteist.autoimageslider.SliderAnimations;
import com.smarteist.autoimageslider.SliderView;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainFragment extends Fragment {
    SliderView sliderView;
    private MainImageSlider adapter;
    private MainSongAdapter adapter_main, adapter_main2, adapter_main3;
    private RecyclerView recycler_first, recycler_second, recycler_third;
    private Call<List<RecommendList>> listmusic;
    private List<RecommendList> list_musics;
    private TextView genre1, genre2, genre3;
    private MainViewModel model;
    private NavController navController;
    public MainFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    private void onItemClick(int index, int position, PlayList playList){
        model.playList.setValue(playList);
        navController.navigate(R.id.action_navigation_home_to_songListFragment);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);


        model = new ViewModelProvider(requireActivity()).get(MainViewModel.class);
        navController = Navigation.findNavController(view);

        recycler_first = view.findViewById(R.id.recycler_first);
        recycler_second = view.findViewById(R.id.recycler_second);
        recycler_third = view.findViewById(R.id.recycler_third);

        genre1 = view.findViewById(R.id.genre1);
        genre2 = view.findViewById(R.id.genre2);
        genre3 = view.findViewById(R.id.genre3);

        adapter_main = new MainSongAdapter(getContext(),((position, playList) -> onItemClick(1, position, playList)));
        adapter_main2 = new MainSongAdapter(getContext(), ((position, playList) -> onItemClick(2, position, playList)));
        adapter_main3 = new MainSongAdapter(getContext(), ((position, playList) -> onItemClick(3, position, playList)));



        recycler_first.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        recycler_second.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));
        recycler_third.setLayoutManager(new LinearLayoutManager(getContext(), LinearLayoutManager.HORIZONTAL, false));

        recycler_first.setAdapter(adapter_main);
        recycler_second.setAdapter(adapter_main);
        recycler_third.setAdapter(adapter_main);

        listmusic = Server.getInstance().getApi().getList();

        listmusic.enqueue(new Callback<List<RecommendList>>() {

            @Override
            public void onResponse(Call<List<RecommendList>> call, Response<List<RecommendList>> response) {
                if(response.code() == 200){
                    list_musics = response.body();
                    adapter_main.setData(list_musics.get(0).getList());
                    adapter_main2.setData(list_musics.get(1).getList());
                    adapter_main3.setData(list_musics.get(2).getList());

                    genre1.setText("#"+list_musics.get(0).getGenreName());
                    genre2.setText("#"+list_musics.get(1).getGenreName());
                    genre3.setText("#"+list_musics.get(2).getGenreName());

                } else if(response.code() == 500) {
                    Toast.makeText(requireContext(), "서버오류", Toast.LENGTH_SHORT).show();
                }
            }

            @Override
            public void onFailure(Call<List<RecommendList>> call, Throwable t) {

            }
        });

        sliderView = view.findViewById(R.id.imageSlider);
        adapter = new MainImageSlider(getContext());

        addNewItem();
        sliderView.setSliderAdapter(adapter);
        sliderView.setIndicatorAnimation(IndicatorAnimationType.WORM); //set indicator animation by using SliderLayout.IndicatorAnimations. :WORM or THIN_WORM or COLOR or DROP or FILL or NONE or SCALE or SCALE_DOWN or SLIDE and SWAP!!
        sliderView.setSliderTransformAnimation(SliderAnimations.SIMPLETRANSFORMATION);
        sliderView.setAutoCycleDirection(SliderView.AUTO_CYCLE_DIRECTION_BACK_AND_FORTH);
        sliderView.setIndicatorSelectedColor(Color.WHITE);
        sliderView.setIndicatorUnselectedColor(Color.GRAY);
        sliderView.setScrollTimeInSec(3);
        sliderView.setAutoCycle(true);
        sliderView.startAutoCycle();

    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_home, container, false);


        return root;
    }
    public void addNewItem() {
        for(int i = 0 ; i < 4; i ++) {
            SliderItem sliderItem = new SliderItem();
            sliderItem.setImageUrl(R.drawable.ic_launcher_background);
            adapter.addItem(sliderItem);
        }
    }
}

File: Wave_V2/app/src/main/java/com/project/wave_v2/view/fragment/MainFragment.kt
package com.project.wave_v2.view.fragment

import android.content.Context.MODE_PRIVATE
import android.content.SharedPreferences
import android.graphics.Color
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.navigation.NavController
import androidx.navigation.Navigation
import androidx.recyclerview.widget.LinearLayoutManager
import com.project.wave_v2.R
import com.project.wave_v2.data.SliderItem
import com.project.wave_v2.data.request.playlist.CallPlayListBody
import com.project.wave_v2.data.response.playlist.CallPlayListModel
import com.project.wave_v2.data.response.playlist.ListInfo
import com.project.wave_v2.network.RetrofitClient
import com.project.wave_v2.network.Service
import com.project.wave_v2.widget.MainAdapter
import com.project.wave_v2.widget.MainImageSlider
import com.smarteist.autoimageslider.IndicatorView.animation.type.IndicatorAnimationType
import com.smarteist.autoimageslider.SliderAnimations
import com.smarteist.autoimageslider.SliderView
import kotlinx.android.synthetic.main.fragment_home.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit


class MainFragment : Fragment() {

    var API : Service? = null
    lateinit var retrofit : Retrofit

    lateinit var navController: NavController

    var firstList : ArrayList<ListInfo> = ArrayList<ListInfo>()
    var secondList : ArrayList<ListInfo> = ArrayList<ListInfo>()
    var thirdList : ArrayList<ListInfo> = ArrayList<ListInfo>()

    var firstAdapter: MainAdapter = MainAdapter(firstList)
    var secondAdapter: MainAdapter = MainAdapter(secondList)
    var thirdAdapter: MainAdapter = MainAdapter(thirdList)

    lateinit var slideAdapter : MainImageSlider

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        retrofit = RetrofitClient.getInstance()
        API = RetrofitClient.getService()

        navController = Navigation.findNavController(view)

        rcViewSetting()

        val prefs: SharedPreferences = requireActivity().getSharedPreferences("user_info", MODE_PRIVATE)
        var id : String? = prefs.getString("userId", "user")


        Log.d("ididid", id)
        API?.getList(
                CallPlayListBody(
                        userId = id
                )
        )?.enqueue(object : Callback<List<CallPlayListModel>> {
            override fun onResponse(call: Call<List<CallPlayListModel>>, response: Response<List<CallPlayListModel>>) {
                if (response.code() == 200) {
                    firstList.clear()
                    secondList.clear()
                    thirdList.clear()

                    for(i in 0 until response.body()?.get(0)?.list?.size!! ){
                        firstList.add(response.body()?.get(0)?.list!![i])
                    }
                    firstAdapter.notifyDataSetChanged()
                    for(i in 0 until response.body()?.get(1)?.list?.size!! ){
                        secondList.add(response.body()?.get(1)?.list!![i])
                    }
                    secondAdapter.notifyDataSetChanged()
                    for(i in 0 until response.body()?.get(2)?.list?.size!! ){
                        thirdList.add(response.body()?.get(2)?.list!![i])
                    }
                    thirdAdapter.notifyDataSetChanged()


                    genre1.text = "#" + response.body()?.get(0)?.genreName
                    genre2.text = "#" + response.body()?.get(1)?.genreName
                    genre3.text = "#" + response.body()?.get(2)?.genreName


                } else {
                    Log.d("Main List", response.code().toString())
                }
            }

            override fun onFailure(call: Call<List<CallPlayListModel>>, t: Throwable) {

            }
        })

        slideAdapter = MainImageSlider(context)
        addNewItem()
        imageSlider.setSliderAdapter(slideAdapter)
        imageSlider.setIndicatorAnimation(IndicatorAnimationType.WORM)
        imageSlider.setSliderTransformAnimation(SliderAnimations.SIMPLETRANSFORMATION)
        imageSlider.autoCycleDirection = SliderView.AUTO_CYCLE_DIRECTION_BACK_AND_FORTH
        imageSlider.indicatorSelectedColor = Color.WHITE
        imageSlider.indicatorUnselectedColor = Color.GRAY
        imageSlider.scrollTimeInSec = 3
        imageSlider.isAutoCycle = true
        imageSlider.startAutoCycle()

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_home, container, false)
    }

    private fun rcViewSetting(){
        recycler_first.adapter = firstAdapter
        recycler_second.adapter = secondAdapter
        recycler_third.adapter = thirdAdapter

        recycler_first.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        recycler_second.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)
        recycler_third.layoutManager = LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false)

        recycler_first.setHasFixedSize(true)
        recycler_second.setHasFixedSize(true)
        recycler_third.setHasFixedSize(true)
    }

    private fun addNewItem() {
        val sliderItem1 = SliderItem()
        val sliderItem2 = SliderItem()
        val sliderItem3 = SliderItem()
        val sliderItem4 = SliderItem()

        sliderItem1.ImageUrl = R.drawable.slide1
        sliderItem2.ImageUrl = R.drawable.slide2
        sliderItem3.ImageUrl = R.drawable.slide3
        sliderItem4.ImageUrl = R.drawable.slide4

        slideAdapter.addItem(sliderItem1)
        slideAdapter.addItem(sliderItem2)
        slideAdapter.addItem(sliderItem3)
        slideAdapter.addItem(sliderItem4)
    }
}